### ISSUE TITLE
关于法线正则化的一些问题

### ISSUE THREAD

**Issue by DDDontbenoob** [hbb1/2d-gaussian-splatting#155]
非常感谢大佬的工作！拜读了论文和代码后关于法线正则化代码这里我有一些问题，问题可能有点仔细，恳请大佬答疑解惑：

1. 根据论文中说的法线正则化我实际推导出来代码应该写成这样：
`normal_error = (rend_alpha - (rend_normal * surf_normal).sum(dim=0))[None] #surf_noamal不乘rend_alpha`
而实际代码中是这样：
`normal_error = (1 - (rend_normal * surf_normal).sum(dim=0))[None] #surf_noamal乘上了rend_alpha`
看起来实际代码是将渲染的法线和估计的表面法线对齐（但实际上rend_normal不是单位向量，surf_normal因为乘了rend_alpha也不是单位向量），那么我想问为什么代码中选择了和论文中不同的实现，也许是两种方法结果差别不大？
2. rend_normal正如代码中注释所说并没有归一化，而是光线路径中每一个splat法线的加权和，正因为rend_normal没有归一化，所以surf_normal最后乘了一个rend_alpha。这里我有疑惑，既然是将渲染法线和表面法线对齐，为什么不直接归一化rend_normal，反而要改变surf_normal？
不知道我有没有表述清楚，或许我在疑惑不重要的细节o(╥﹏╥)o

**Comment by hbb1** [comment]
实际上在原本的损失上需要乘以 accumulated alpha 来降低透明区域的损失，这样会对半透明的场景比如blender 比较友好。
```python
render_alpha.detach() * (1 - rend_norml  @ unnomarlized_surf_normal)
= render_alpha.detach() - render_normal @ (render_alpha.detach() * unnomarlized_surf_normal)
= render_alpha.detach() - rend_normal @ surf_normal
```
由于是detach 的所以 render_alpha 就无所谓了。

**Comment by DDDontbenoob** [comment]
> 实际上在原本的损失上需要乘以 accumulated alpha 来降低透明区域的损失，这样会对半透明的场景比如blender 比较友好。
> 
> ```python
> render_alpha.detach() * (1 - rend_norml  @ unnomarlized_surf_normal)
> = render_alpha.detach() - render_normal @ (render_alpha.detach() * unnomarlized_surf_normal)
> = render_alpha.detach() - rend_normal @ surf_normal
> ```
> 
> 由于是detach 的所以 render_alpha 就无所谓了。

懂了！感谢大佬~
